pkg/broker/handler/handler_test.go

```diff
@@ -141,6 +138,91 @@ func TestHandler(t *testing.T) {
 	})
 }
 
+type alwaysErrProc struct {
+	processors.BaseProcessor
+	desiredErrCount, currErrCount int
+}
+
+func (p *alwaysErrProc) Process(_ context.Context, _ *event.Event) error {
+	if p.currErrCount < p.desiredErrCount {
+		p.currErrCount++
+		return errors.New("always error")
+	}
+	return nil
+}
+
+func TestRetryBackoff(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	c, close := testPubsubClient(ctx, t, "test-project")
+	defer close()
+
+	topic, err := c.CreateTopic(ctx, "test-topic")
+	if err != nil {
+		t.Fatalf("failed to create topic: %v", err)
+	}
+	sub, err := c.CreateSubscription(ctx, "test-sub", pubsub.SubscriptionConfig{
+		Topic: topic,
+	})
+	if err != nil {
+		t.Fatalf("failed to create subscription: %v", err)
+	}
+
+	p, err := cepubsub.New(context.Background(),
+		cepubsub.WithClient(c),
+		cepubsub.WithProjectID("test-project"),
+		cepubsub.WithTopicID("test-topic"),
+	)
+	if err != nil {
+		t.Fatalf("failed to create cloudevents pubsub protocol: %v", err)
+	}
+
+	delays := []time.Duration{}
+	desiredErrCount := 8
+	processor := &alwaysErrProc{desiredErrCount: desiredErrCount}
+	h := NewHandler(sub, processor, time.Second, RetryPolicy{MinBackoff: time.Millisecond, MaxBackoff: 16 * time.Millisecond})
+	// Mock sleep func to collect nack backoffs.
+	h.delayNack = func(d time.Duration) {
+		delays = append(delays, d)
+	}
+	h.Start(ctx, func(err error) {})
+	defer h.Stop()
+	if !h.IsAlive() {
+		t.Error("start handler didn't bring it alive")
+	}
+
+	testEvent := event.New()
+	testEvent.SetID("id")
+	testEvent.SetSource("source")
+	testEvent.SetSubject("subject")
+	testEvent.SetType("type")
+
+	if err := p.Send(ctx, binding.ToMessage(&testEvent)); err != nil {
+		t.Errorf("failed to seed event to pubsub: %v", err)
+	}
+
+	// Wait until all desired errors were returned.
+	// Then stop the handler by cancel the context.
+	time.Sleep(time.Second)
+	cancel()
+
+	if len(delays) != desiredErrCount {
+		t.Errorf("retry count got=%d, want=%d", len(delays), desiredErrCount)
+	}
+	if delays[0] != time.Millisecond {
+		t.Errorf("initial nack delay got=%v, want=%v", delays[0], time.Millisecond)
+	}
+	// We expect exponential backoff until MaxBackoff
+	for i := 1; i < len(delays); i++ {
+		wantDelay := 2 * delays[i-1]
+		if wantDelay > 16*time.Millisecond {
```

Comment:
this seems unnecessary. Looks like it can only happen when len(delays)>desiredErrCount, but we already have an `Errorf`, we can even make it `Fatalf`

Reply:
This is necessary because I intentionally had enough errors to reach max delay more than once. This is to ensure the code never delays more than the max delay.